### ISSUE TITLE
RFC: Store flattened nested form state internally with dotpath: value

### ISSUE THREAD

**Issue by DavidBabel** [jaredpalmer/formik#1103]
## 🚀 Feature request

Speed up Formik by preventing the use of **toPath** / **getIn** / **setIn** where it's not necessary.
It' simple, but not fast to do.

### Current Behavior

Each of this methods : **toPath** (lodash) / **getIn** (internal) / **setIn** (internal) are very consuming. With 50 fields in a simple form, if i type only 3 characters in a `FastField`, i got close to 12000 loop launched (!) only for the "shouldComponentUpdate" part.
![image](https://user-images.githubusercontent.com/5599375/48346480-916ce980-e67b-11e8-8060-ea53ac3f0805.png)

The problem starts with a good idea. Use the **_.toPath** (which is handfull with the result, but only for the result actually). But [toPath](https://github.com/lodash/lodash/blob/master/toPath.js) use two loops [here](https://github.com/lodash/lodash/blob/master/map.js) and [here](https://github.com/lodash/lodash/blob/master/.internal/copyArray.js) and should be used only when needed.

So the problem appear because of this implementation stategy, way too much loop are triggered on rerendering, making the whole app laggy.

### Desired Behavior

Same behaviour for user, but less laggy.

### Suggested Solution

The solution is actually quite simple. instead of managing a multi level object inside the Formik internal state (thanks to the "toPath"), why not only store the full object path as a key ?

It will be very easier to implement the **getIn** / **setIn** without any loop, and save the use of **toPath** only for the submitted result.

Replace library internal state from : 
```json
{
  "person" : {
    "name": "myname",
    "forname": "myforname"
  }
}
```

by : 
```json
{
  "person.name" : "myname",
  "person.forname": "myforname"
}
```
And **only** us a **toPath** : 
- onSubmit()
- setTouched()
- set...Anything() // the provided methods of formik

This will save a LOT of unneeded calculation loops, for `Field` and `FastField` regardless.

### Who does this impact? Who is this for?

It can be transparent for users.
Formik can become scalable.
It can save this library and the world :)

### Describe alternatives you've considered

I made a very quick change to the `FastFields` to only render if they are the last touched field only : 
here is the PR : https://github.com/jaredpalmer/formik/pull/1104
implemented here for test purposes : https://www.npmjs.com/package/@davidbabel/formik

Just implemented a "lastTouched" item which contain the name of the last touched item. And it save a looooot of loops during rendering : 
![image](https://user-images.githubusercontent.com/5599375/48352284-ad2cbb80-e68c-11e8-84d3-06b158b5d28a.png)

**Note** that both of the approaches (this issue and lastTouched) are huge upgrade in term of performances.

### Additional context

In any `<form>` with more than 15 fields, it become very cpu consuming to use Formik in any use case. And it's an easy workaround.
It may be a breaking change for some usage of formik it don't understand now.

**Comment by DavidBabel** [comment]
I made a very quick change to the `FastFields` to only render if they are the last touched field only : 
here is the PR : https://github.com/jaredpalmer/formik/pull/1104

Just implemented a "lastTouched" item which contain the name of the last touched item.

**Comment by jaredpalmer** [comment]
Yes this would help. However, it’s not as easy it looks. You’ll have to rewrite FieldArray and also some internals as well. Also, you’ll have to flatten/unflatten in other places as well.

**Comment by jaredpalmer** [comment]
To maintain backwards compat you’ll also have unflatten inside of every render for values, errors, touched.

**Comment by DavidBabel** [comment]
Yeah i know it's a "big" thing. That's why i did not jump into it directly, this ecosystem is far too complex to be able to mesure every possible issues.
But i'm pretty sure that the performances will increase drastically with this.

**Comment by FredyC** [comment]
I would like to add, that we should be able to specify the nested field name as an array and in that case, it would be possible to skip over lodash shenanigans to improve performance. Technically, it works now already, because `_.toPath` accepts an array (even though it's not documented), but does [extra sanity checks](https://github.com/lodash/lodash/blob/master/toPath.js#L23) for items in an array which we don't need I think.

**Comment by jaredpalmer** [comment]
I think this is actually a huge perf win we should re-explore.
